### PR TITLE
New featured response functionality, bug fixes, changes for cmd deliverable

### PR DESCRIPTION
**Pull request by hknehans**
To make a deliverable work, I had to rebuild the sentiment model with tensorflow 2.7 and add a couple new arguments for saving and loading. There are also a couple bug fixes for the sentiment analysis formatting. New functionality to find and show top responses showcasing average sentiment.

### CODE REVIEW

File: analysis/Sentiment_Analysis.py [hunk 1]
```diff
@@ -1,18 +1,22 @@
-from analysis.analysis import Analysis
+from analysis.Analysis import Analysis
 import os
-import tensorflow.keras
-# TODO this can be removed once the word_index is saved and loaded in
-from tensorflow.keras.datasets import imdb
+os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2' # suppresses tf info and warning logs
```
Comment (hknehans) [inline analysis/Sentiment_Analysis.py:3]: I had to change the way we were importing some of this to make the cmd deliverable work

File: analysis/Sentiment_Analysis.py [hunk 2]
```diff
@@ -1,18 +1,22 @@
...
+from tensorflow.keras.datasets.imdb import get_word_index
 from tensorflow.keras.preprocessing.text import text_to_word_sequence
-from tensorflow.keras.preprocessing import sequence
-
+from tensorflow.keras.preprocessing.sequence import pad_sequences
 
 class Sentiment_Analysis(Analysis):
+    
+    # model_path = os.path.join(os.path.dirname(__file__), 'sentiment_model.tf')
+    model_path = os.path.join(os.path.expanduser('~'), 'FRAT-models\\2_7_sentiment_model.tf')
```
Comment (hknehans) [inline analysis/Sentiment_Analysis.py:13]: `os.path.expanduser('~')` is equivalent to your home path. `model_path` should be something similar to 'C:\Users\yourUser\FRAT-models\2_7_sentiment_model.tf'

File: analysis/Sentiment_Analysis.py [hunk 3]
```diff
@@ -63,63 +68,192 @@ def analyze(cls, padded_sequences):
...
+        average_sentiment_type = ""
         negatives_count = neutrals_count = positives_count = 0
 
-        # Iterate through to get sentiment sum and the counts of each sentiment type
+        sum = 0
+        # Iterate through to get the counts of each sentiment type
         for sentiment in sentiments:
-            sentiment_value = sentiment[0]
-            sum += sentiment_value
+            sentiment = round(sentiment[0].item(), 2)
```
Comment (hknehans) [inline analysis/Sentiment_Analysis.py:86]: fixes previous bug

File: analysis/Sentiment_Analysis.py [hunk 4]
```diff
@@ -63,63 +68,192 @@ def analyze(cls, padded_sequences):
...
 
-        # TODO percentage rounding is not consistent. Sometimes will have lots of decimal points
-        sentiment_analysis_results["overall"] = round(sum / len(sentiments), 2)
-        sentiment_analysis_results["percent_negative"] = round(
-            (negatives_count / len(sentiments)), 2) * 100
-        sentiment_analysis_results["percent_neutral"] = round(
-            (neutrals_count / len(sentiments)), 2) * 100
-        sentiment_analysis_results["percent_positive"] = round(
-            (positives_count / len(sentiments)), 2) * 100
+        sentiment_analysis_results["percent_negative"] = round((negatives_count / len(sentiments)) * 100)
```
Comment (hknehans) [inline analysis/Sentiment_Analysis.py:96]: fixes previous rounding bug

File: analysis/Thematic_Anlaysis.py [hunk 5]
```diff
@@ -1,13 +1,11 @@
 from analysis.Analysis import Analysis
-import nltk # TODO this can be removed when nltk.download('stopwords') is saved and loaded in
 from nltk.corpus import stopwords
 from textblob import Word, TextBlob
 from sklearn.decomposition import NMF
 from sklearn.feature_extraction.text import TfidfVectorizer
 
 class Thematic_Analysis(Analysis):
 
-    nltk.download('stopwords') # TODO this may need to be saved somewhere too so we can load it in locally without having to download it
```
Comment (hknehans) [inline analysis/Thematic_Anlaysis.py:10]: this is under the belief that `from nltk.corpus import stopwords` is enough

File: analysis/build_sentiment_model.ipynb [hunk 6]
```diff
@@ -0,0 +1,225 @@
+{
```
Comment (hknehans) [inline analysis/build_sentiment_model.ipynb:1]: just a renaming of `drew-model.ipynb` and some small changes for tf:2.7

File: user_input/input.py [hunk 7]
```diff
@@ -11,20 +11,14 @@
 ===================================================================
 """
 def response_option(max_len):
-    valid_response = False
-    while not valid_response:
-        response = str(input("Enter in a response with no more than {} words: ".format(max_len)))
-        if len(response.split()) > max_len:
```
Comment (hknehans) [inline user_input/input.py:17]: we are just going to truncate the response in `pad_sequence()` if more than `max_len` is inputted

File: user_input/input.py [hunk 8]
```diff
@@ -64,16 +60,14 @@ def bom_validation(filename, default='utf-8'):
...
-    with open(path, newline='', encoding=encodingX) as csv_file:
-        csv_reader = csv.reader(csv_file)
+
+    with open(path, newline='', encoding=encodingX) as csv_file:    
+        csv_reader = csv.reader(csv_file, delimiter='|')
         for row in csv_reader:
-            # TODO Cleanup if/else and try/except
-            # Validate that there is only one element in each record.
-            try:
-                responses.append(row[0])
```
Comment (hknehans) [inline user_input/input.py:74]: assuming only one response per row is inputted

**Review by DrewRinker — APPROVED** [review]

**Pull request merged**